File: stackunderflow/etl/normalize/openclaw.py
"""OpenClaw normalizer.

OpenClaw (and the related ``clawdbot`` / ``moltbot`` / ``moldbot``
agents that share the format) writes JSONL sessions where each
assistant message carries an explicit ``usage`` block:

    { "type": "message",
      "message": {
        "role": "assistant",
        "content": [...],
        "model": "...",
        "provider": "...",
        "usage": {
          "input": ...,
          "output": ...,
          "cacheRead": ...,
          "cacheWrite": ...,
          "cost": {"total": ...}   # optional, provider-embedded
        }
      } }

Policy:

* Trust the explicit token block; map ``cacheWrite`` →
  ``cache_create_tokens``, ``cacheRead`` → ``cache_read_tokens``.
* Stamp ``cost_source='rate_card'`` for known model ids /
  ``'unknown'`` otherwise. (We recompute via ``compute_cost`` for
  parity with the rest of the pipeline; the embedded ``cost.total``
  is preserved in ``raw_extras`` as a cross-reference but not
  consumed.)
"""

from __future__ import annotations

import json
from collections.abc import Iterable

from stackunderflow.infra.costs import RATE_CARD

from .base import (
    COST_SOURCE_RATE_CARD,
    COST_SOURCE_UNKNOWN,
    Normalizer,
)

_DEFAULT_MODEL = "openclaw-auto"
_RAW_EXTRAS_FIELDS = ("provider", "agentName", "embeddedCost")


class OpenClawNormalizer(Normalizer):
    provider_name = "openclaw"

    def normalize(self, msg_row: dict) -> Iterable[dict]:
        role = str(msg_row.get("role") or "")
        if role != "assistant":
            return

        canonical = _canonical_tokens(msg_row)
        if canonical is None:
            return

        if all(canonical[k] == 0 for k in ("input", "output", "cache_read", "cache_create")):
            return

        model = str(msg_row.get("model") or "") or _DEFAULT_MODEL

        cost_source = (
            COST_SOURCE_RATE_CARD
            if model in RATE_CARD
            else COST_SOURCE_UNKNOWN
        )

        raw_extras = _extras_from_raw_json(msg_row.get("raw_json"))

        yield self._build_event(
            msg_row,
            input_tokens=canonical["input"],
            output_tokens=canonical["output"],
            cache_read_tokens=canonical["cache_read"],
            cache_create_tokens=canonical["cache_create"],
            cost_source=cost_source,
            model=model,
            raw_extras=raw_extras,
        )


def _canonical_tokens(msg_row: dict) -> dict[str, int] | None:
    """Return canonical 4-key shape from the OpenClaw usage block."""
    # Prefer raw usage block (production path); fall back to columns.
    raw = _raw_usage(msg_row)
    if raw is not None:
        return {
            "input": _safe_int(raw.get("input")),
            "output": _safe_int(raw.get("output")),
            "cache_read": _safe_int(raw.get("cacheRead")),
            "cache_create": _safe_int(raw.get("cacheWrite")),
        }
    return {
        "input": int(msg_row.get("input_tokens") or 0),
        "output": int(msg_row.get("output_tokens") or 0),
        "cache_read": int(msg_row.get("cache_read_tokens") or 0),
        "cache_create": int(msg_row.get("cache_create_tokens") or 0),
    }


def _raw_usage(msg_row: dict) -> dict | None:
    """Pull the OpenClaw ``usage`` block from msg_row or raw_json."""
    direct = msg_row.get("usage")
    if isinstance(direct, dict):
        return direct
    payload = _safe_load_raw(msg_row.get("raw_json"))
    if isinstance(payload, dict):
        # The persisted shape may have ``message.usage`` (full event)
        # or ``usage`` directly (already unwrapped).
        msg = payload.get("message")
        if isinstance(msg, dict):
            usage = msg.get("usage")
            if isinstance(usage, dict):
                return usage
        usage = payload.get("usage")
        if isinstance(usage, dict):
            return usage
    return None


def _safe_int(value: object) -> int:
    try:
        return max(int(value or 0), 0)
    except (TypeError, ValueError):
        return 0


def _extras_from_raw_json(raw_json: object) -> dict | None:
    payload = _safe_load_raw(raw_json)
    if not isinstance(payload, dict):
        return None
    out: dict = {}
    inner = payload.get("message") if isinstance(payload.get("message"), dict) else payload
    for key in _RAW_EXTRAS_FIELDS:
        val = inner.get(key) if isinstance(inner, dict) else None
        if val is not None and val != "":
            out[key] = val
    # Preserve embedded cost for cross-reference. Inspect the parsed
    # payload directly — ``_raw_usage`` operates on a msg_row shape
    # (looking at ``raw_json`` key) which won't be present here.
    usage = None
    if isinstance(inner, dict) and isinstance(inner.get("usage"), dict):
        usage = inner["usage"]
    elif isinstance(payload.get("usage"), dict):
        usage = payload["usage"]
    if isinstance(usage, dict):
        cost = usage.get("cost")
        if cost is not None:
            out["embeddedCost"] = cost
    return out or None


def _safe_load_raw(raw_json: object) -> object | None:
    if isinstance(raw_json, dict):
        return raw_json
    if not isinstance(raw_json, str | bytes | bytearray):
        return None
    try:
        if isinstance(raw_json, bytes | bytearray):
            return json.loads(raw_json.decode("utf-8", errors="replace"))
        return json.loads(raw_json)
    except (json.JSONDecodeError, ValueError, UnicodeDecodeError):
        return None

File: rust/crates/stax-etl/src/normalize/openclaw.rs
//! OpenClaw — port of `etl/normalize/openclaw.py`.
//!
//! Same explicit-`usage`-block shape as hermes and pi, with one addition: the
//! block may carry a provider-embedded `cost`, which is preserved in
//! `raw_extras` as `embeddedCost` for cross-reference and **not** consumed —
//! every row re-prices through `compute_cost` so the marts read one number
//! computed one way.

use stax_core::queries::pyjson::Value as PyValue;

use super::base::{EventSpec, NormalizeContext, Normalizer, UsageEvent};
use super::row::{MsgRow, PyRaise, as_dict, int_or_zero, safe_int, safe_load_raw, str_or_empty};
use super::text::{collect_fields, unwrap_envelope};
use crate::pricing::Tokens;

const DEFAULT_MODEL: &str = "openclaw-auto";
const RAW_EXTRAS_FIELDS: [&str; 3] = ["provider", "agentName", "embeddedCost"];

/// The `openclaw` normalizer.
#[derive(Debug, Clone, Copy, Default)]
pub struct OpenClawNormalizer;

impl Normalizer for OpenClawNormalizer {
    fn provider_name(&self) -> &'static str {
        "openclaw"
    }

    fn normalize(&self, ctx: &NormalizeContext, row: &MsgRow) -> Result<Vec<UsageEvent>, PyRaise> {
        if str_or_empty(row, "role") != "assistant" {
            return Ok(Vec::new());
        }

        let canonical = canonical_tokens(row)?;
        if canonical.input == 0
            && canonical.output == 0
            && canonical.cache_read == 0
            && canonical.cache_creation == 0
        {
            return Ok(Vec::new());
        }

        let model = match str_or_empty(row, "model") {
            empty if empty.is_empty() => DEFAULT_MODEL.to_string(),
            model => model,
        };
        let cost_source = ctx.rate_card_or_unknown(&model);

        Ok(vec![
            self.build_event(
                ctx,
                row,
                EventSpec::new(
                    canonical.input,
                    canonical.output,
                    canonical.cache_read,
                    canonical.cache_creation,
                    cost_source,
                )
                .model(model)
                .raw_extras(extras_from_raw_json(row.get("raw_json"))),
            ),
        ])
    }
}

fn canonical_tokens(row: &MsgRow) -> Result<Tokens, PyRaise> {
    if let Some(usage) = raw_usage(row) {
        return Ok(Tokens {
            input: safe_int(usage.get("input"))?,
            output: safe_int(usage.get("output"))?,
            cache_read: safe_int(usage.get("cacheRead"))?,
            cache_creation: safe_int(usage.get("cacheWrite"))?,
        });
    }
    Ok(Tokens {
        input: int_or_zero(row, "input_tokens")?,
        output: int_or_zero(row, "output_tokens")?,
        cache_read: int_or_zero(row, "cache_read_tokens")?,
        cache_creation: int_or_zero(row, "cache_create_tokens")?,
    })
}

fn raw_usage(row: &MsgRow) -> Option<PyValue> {
    if let Some(direct) = as_dict(row.get("usage")) {
        return Some(direct.clone());
    }
    let payload = safe_load_raw(row.get("raw_json"))?;
    if !matches!(payload, PyValue::Object(_)) {
        return None;
    }
    if let Some(message) = as_dict(payload.get("message"))
        && let Some(usage) = as_dict(message.get("usage"))
    {
        return Some(usage.clone());
    }
    as_dict(payload.get("usage")).cloned()
}

/// The keepsake sweep plus the embedded cost.
///
/// The `usage` block is looked up on the *parsed payload* directly rather than
/// through [`raw_usage`], which expects a `msg_row` shape — Python leaves a
/// comment saying exactly that, because the two lookups are one keystroke apart
/// and the wrong one silently finds nothing.
fn extras_from_raw_json(raw_json: Option<&PyValue>) -> Option<PyValue> {
    let payload = safe_load_raw(raw_json)?;
    if !matches!(payload, PyValue::Object(_)) {
        return None;
    }
    let inner = unwrap_envelope(&payload, "message");
    let mut out = collect_fields(inner, &RAW_EXTRAS_FIELDS);

    let usage = as_dict(inner.get("usage")).or_else(|| as_dict(payload.get("usage")));
    if let Some(usage) = usage
        && let Some(cost) = usage.get("cost")
        && !matches!(cost, PyValue::Null)
    {
        // `out["embeddedCost"] = cost` — an assignment, so a key already
        // collected by the sweep is REPLACED in place rather than skipped.
        match out.iter_mut().find(|(name, _)| name == "embeddedCost") {
            Some(slot) => slot.1 = cost.clone(),
            None => out.push(("embeddedCost".to_string(), cost.clone())),
        }
    }
    (!out.is_empty()).then_some(PyValue::Object(out))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::normalize::base::CostSource;
    use crate::normalize::test_support::{assistant_row, ctx};

    fn openclaw_row(usage: &str) -> MsgRow {
        assistant_row("openclaw", "claude-sonnet-4-5-20250929").with(
            "raw_json",
            PyValue::Str(format!(
                r#"{{"message": {{"usage": {usage}, "agentName": "claw"}}}}"#
            )),
        )
    }

    #[test]
    fn the_embedded_cost_is_preserved_and_not_consumed() {
        let row = openclaw_row(
            r#"{"input": 1000, "output": 100, "cacheRead": 5, "cacheWrite": 6,
                "cost": {"total": 99.0}}"#,
        );
        let events = OpenClawNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!(events[0].input_tokens, 1000);
        assert_eq!(events[0].cache_create_tokens, 6);
        assert_eq!(
            events[0].raw_extras.as_deref(),
            Some(r#"{"agentName": "claw", "embeddedCost": {"total": 99.0}}"#)
        );
        // Re-priced, not taken from the payload.
        assert!(events[0].cost_usd > 0.0);
        assert!(events[0].cost_usd < 1.0);
        assert_eq!(events[0].cost_source, CostSource::RateCard);
    }

    #[test]
    fn garbage_usage_values_degrade_to_zero_and_can_drop_the_row() {
        let row = openclaw_row(r#"{"input": "abc", "output": null}"#);
        assert!(
            OpenClawNormalizer
                .normalize(&ctx(), &row)
                .unwrap()
                .is_empty()
        );
    }
}
